Add `Builtin.isConcrete<T>(T.Type) -> Int1`

nvzqz opened the pull request:
Returns `true` if `T.Type` is known to refer to a concrete type during IRGen. Otherwise, if still generic, then it is lowered to `false` right before IRGen in IRGenPrepare. This allows for the optimizer to specialize this at -O and eliminate conditional code.

This also Includes:
- `Swift._isConcrete<T>(T.Type) -> Bool` wrapper function.
- tests on the [un]optimized SIL generated from calling `isConcrete` on generic and concrete types in -O[none] builds.

This is prior work to adding `Builtin.generic_add<T>(T, T) -> T` and friends for SIMD vector types.

nvzqz committed:
Add `Builtin.isConcrete<T>(T.Type) -> Int1`

Returns `true` if `T.Type` is known to refer to a concrete type. The
implementation allows for the optimizer to specialize this at -O and
eliminate conditional code.

Includes `Swift._isConcrete<T>(T.Type) -> Bool` wrapper function.

stephentyrone commented:
Thanks @nvzqz! I'll take a look later today.

stephentyrone commented:
@swift-ci please smoke test

jckarter commented:
Our existing precedent for this is to name these sorts of things `canBe`. Check out `canBeClass` for an example. `canBeClass` also returns a tristate, because it ended up being valuable to know whether the answer is "definitely no", "definitely yes", or "maybe".

stephentyrone commented:
> Our existing precedent for this is to name these sorts of things `canBe`. Check out `canBeClass` for an example. `canBeClass` also returns a tristate, because it ended up being valuable to know whether the answer is "definitely no", "definitely yes", or "maybe".

Is this actually tri-state logically though, or do we always know by the time we reach IRGen?

jckarter commented:
If you look at how `TypeBase::canBeClass` is implemented in the `SILCombinerBuiltinVisitors`, on top of `TypeTraitResult TypeBase::canBeClass()`, there should also be a bunch of existing code you can factor out and share between these two traits.

nvzqz commented:
@jckarter that name does seem better. I'm not sure if there's a way of expressing this as a tristate.

jckarter commented:
@nvzqz Maybe not in this case, but I think by following the pattern set by `canBeClass`, you could leverage a bunch of existing implementation, and make it easier to add more traits like this in the future.

jckarter commented:
The way that `canBeClass` works is that the optimizer immediately constant-folds to `Is` or `IsNot` when the trait method gives a definite answer, but leaves the builtin in place for `CanBe`. IRGen then lowers the builtin to the constant value `CanBe` if it hasn't already been lowered away. In the case of "is-concrete", the only interesting cases are `Is` and `CanBe`, I guess, but I think you still want to share the same approach, because this allows the builtin to constant-fold early, thereby allowing SIL-level optimizations to take advantage of the constant folding, without preventing it from also folding later as more information gets exposed.

nvzqz commented:
@swift-ci please smoke test

stephentyrone commented:
@swift-ci please smoke test

stephentyrone commented on stdlib/public/core/Builtin.swift:
```diff
@@ -691,6 +691,14 @@ func _isPOD<T>(_ type: T.Type) -> Bool {
 }
 
+/// Returns `true` if `type` is known to refer to a concrete type during IRGen.
+@_transparent
```
I'm fairly sure that `@_transparent` is redundant with `@_aEIC` here.

stephentyrone reviewed (approved):
LGTM, modulo @jckarter's questions about naming and reuse of existing code.

nvzqz commented:
After thinking about it some more, I'm not sure if a tri-state naming convention makes sense for this. I think that `isConcrete` already says what it does on the tin in that it expresses how the compiler sees the metatype in a given situation and configuration (e.g. -O and -Onone).

gottesmm commented:
Just as an exploration of the state space, consider the following:

Imagine that we had a builtin called canBeConcrete. It returns either always, never, sometimes (i.e. a tri-state).

If we have a concrete type, easily the builtin must return always since once a type is concrete we can not further specialize it.

What about generic types? Easily, there are two possibilities:

* If we are still allowing for further specialization, we want to return maybe. This ensures that we do not eliminate the builtin too early.

* If we somehow have the knowledge that no matter specialization will occur (for whatever reason), then we can lower to never.

The key thing to notice is that the 2nd bullet point via our semantics is IRGen time/Lowered SIL. That might be the way forward for this in a way that keeps the logic of when to fold this in a utility that the various code can invoke.

Thats all I got.

NOTE: If we were to embed these semantics if we don't assert that the specialization machinery never runs in Lowered SIL, we would necessarily have to do so in order to be careful and prevent mistakes.

jckarter commented:
@nvzqz I commented on the linked commit. You should not be erasing the instruction from inside the `SILCombiner` method, since that will mess up SILCombiner's state. It'll clean up for you. You'd also need to implement the IRGen fallback implementation that emits `false` when the builtin still exists at IRGen time. With those changes, you should get the same benefits as `canBeClass` even if it isn't literally implemented the same. I don't care too much about the naming, but I think having an idiosyncratic name does help call attention to the fact that these have unusual compile-time-state behavior.

nvzqz commented:
I walked through my implementations of @jckarter's suggestions with @atrick and found places where it was ill-formed. The concerns I expressed to @airspeedswift and @gottesmm previously with regard to eager evaluation (always emitting `false`/0) were apparently an artifact of this.

nvzqz commented:
@swift-ci please smoke test

stephentyrone commented on test/stdlib/Builtins.swift:
```diff
@@ -293,4 +293,17 @@ tests.test("_isOptional") {
+  func isConcrete_true<T>(_ type: T.Type) -> Bool {
+    return _isConcrete(type)
+  }
+  func isConcrete_false<T>(_ type: T.Type) -> Bool {
```
What would the effect of marking this `@inline(__always)` be? Does `isConcrete_true` require `@_transparent` specifically? If not, is this fragile to inlining decisions? (and so does `isConcrete_false` need to be marked `@inline(never)`?)

atrick replied:
`@inline(__always)` is *still* not fully implemented--it does nothing at `-Onone`. So `isConcrete_true` requires `@_transparent`. `isConcrete_false` is fragile to inlining decisions. It happens to work here because the test is only run at `-Onone`, but this assumption could be called out explicitly by adding `@inline(never)`.

stephentyrone replied:
Let's please add the `@inline(never)` here, then.

nvzqz replied:
Why don't we test builtins in `-O` on top of `-Onone`?

atrick replied:
Any optimizations that might affect the builtin should definitely have their own tests. You've already done that for constant folding. Now that you've added the SILCombine optimization you should have corresponding test somewhere like `test/SILOptimizer/sil_combine.sil`

stephentyrone replied:
Thanks! I think we just need Andy's suggested test, now.

nvzqz replied:
Added test to bottom of `test/SILOptimizer/sil_combine.sil`.

stephentyrone commented on stdlib/public/core/Builtin.swift:
```diff
@@ -691,6 +691,13 @@ func _isPOD<T>(_ type: T.Type) -> Bool {
   return Bool(Builtin.ispod(type))
 }
 
+/// Returns `true` if `type` is known to refer to a concrete type during IRGen.
```
Can you expand this doc comment with a positive and negative example, please? I know the existing comments are pretty terse, but we don't need to keep doing that.

nvzqz committed:
Never inline `isConcrete_false` test

The `Builtin.isConcrete` function is apparently sensitive to inlining in
unoptimized builds.

stephentyrone commented on lib/SILOptimizer/Utils/ConstantFolding.cpp:
```diff
@@ -1563,7 +1578,8 @@ void ConstantFolder::initializeWorklist(SILFunction &f) {
       }
 
-      if (isApplyOfBuiltin(*inst, BuiltinValueKind::GlobalStringTablePointer)) {
+      if (isApplyOfBuiltin(I, BuiltinValueKind::GlobalStringTablePointer) ||
```
Looks like this should be `*inst` instead of `I`. Did part of the change get dropped somewhere?

nvzqz committed:
Fix variable reference in ConstantFolding phase

nvzqz committed:
Elaborate in docs for `Swift._isConcrete`

stephentyrone commented on stdlib/public/core/Builtin.swift:
```diff
@@ -691,7 +691,12 @@ func _isPOD<T>(_ type: T.Type) -> Bool {
+/// this returns `false` if the check has failed.
+///
+/// Note that there may be cases in which, despite `T` being concrete at some
+/// point in the caller chain, this function will return `false`.
```
Maybe add that it will never return a false positive.

airspeedswift commented:
@swift-ci please smoke test

atrick commented on lib/SILOptimizer/Utils/ConstantFolding.cpp:
```diff
@@ -1572,7 +1588,6 @@ void ConstantFolder::initializeWorklist(SILFunction &f) {
           isa<CheckedCastAddrBranchInst>(inst) ||
           isa<UnconditionalCheckedCastInst>(inst) ||
           isa<UnconditionalCheckedCastAddrInst>(inst)) {
-        WorkList.insert(inst);
```
As you pointed out, this line was mistakenly removed.

atrick commented on lib/SILOptimizer/Utils/ConstantFolding.cpp:
```diff
@@ -1781,6 +1796,17 @@ ConstantFolder::processWorkList() {
+      }
+      continue;
+    }
+
```
Optimizing this builtin works by:
1. Inserting it in the worklist during initializeWorklist, which you've done
2. Adding a call to constantFoldIsConcrete to contentFoldBuiltin, which you've done-- that will be called for everything in the worklist.

So, this extra optimization can just be removed. If I'm right, your tests should still pass.

nvzqz replied:
So I ran my tests against this bit being removed and it starts to fail. So apparently this is a necessary component. 🤷‍♂️

stephentyrone replied:
@atrick gentle ping

atrick replied:
I dunno, but it definitely doesn't make sense to do the same optimization in two places. To debug it take a minimal .sil test case and run `sil-opt -diagnostic-constant-propagation t.sil`. My reading of the code is that your change to `initalizeWorklist` would push the builtin apply instruction on the worklist, and `constantFoldBuiltin` should get a chance to run on each worklist item.

nvzqz commented:
@swift-ci please smoke test

nvzqz commented:
@swift-ci please smoke test

nvzqz commented on test/SILOptimizer/irgenprepare.sil:
```diff
@@ -0,0 +1,31 @@
+// RUN: %target-sil-opt -module-name Swift -irgen-prepare %s -enable-sil-verify-all | %FileCheck %s
```
@gottesmm I imagine this test isn't necessary anymore and we should instead be testing the IRGen phase?

This test is failing when I believe it used to pass before the stuff got moved to IRGen from IRGenPrepare.

gottesmm replied:
Yes.

nvzqz committed:
Remove `Builtin.isConcrete` IRGenPrepare SIL test

nvzqz committed:
Add IRGen SIL test for `Builtin.isConcrete`

nvzqz commented:
@swift-ci please smoke test

nvzqz commented on stdlib/public/core/Builtin.swift:
```diff
@@ -691,6 +691,18 @@ func _isPOD<T>(_ type: T.Type) -> Bool {
+///
+/// Note that there may be cases in which, despite `T` being concrete at some
+/// point in the caller chain, this function will return `false`.
+@_alwaysEmitIntoClient
```
@stephentyrone it seems that the `@_alwaysEmitIntoClient` attribute is causing `_isConcrete` to fail to return `true`. `@_transparent` seems to work so I'll revert it to that. Would it be redundant to keep `@ _alwaysEmitIntoClient` at that point?

stephentyrone replied:
No, you'll need both. `@_transparent` gets you the optimization you want, `@_aEIC` saves you from needing availability.

nvzqz commented on test/IRGen/builtin_isConcrete.sil:
```diff
@@ -0,0 +1,27 @@
+  var value: Builtin.Int32
+}
+
+// CHECK-LABEL: define swiftcc i1 @isConcrete_true() {{.*}} {
```
```
21:53:56 FAIL: Swift(linux-x86_64) :: IRGen/builtin_isConcrete.sil (11076 of 12152)
21:53:56 ******************** TEST 'Swift(linux-x86_64) :: IRGen/builtin_isConcrete.sil' FAILED ********************
21:53:56 Script:
21:53:56 --
21:53:56 : 'RUN: at line 1';   /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift -frontend -target x86_64-unknown-linux-gnu  -module-cache-path '/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache' -swift-version 4   -typo-correction-limit 10  -emit-ir -parse-sil /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/test/IRGen/builtin_isConcrete.sil -module-name Swift -parse-stdlib | '/usr/bin/python' '/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/utils/PathSanitizingFileCheck' --sanitize BUILD_DIR='/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/swift-linux-x86_64' --sanitize SOURCE_DIR='/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift' --use-filecheck '/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/llvm-linux-x86_64/bin/FileCheck'  /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/test/IRGen/builtin_isConcrete.sil
21:53:56 --
21:53:56 Exit Code: 1
21:53:56 
21:53:56 Command Output (stderr):
21:53:56 --
21:53:56 
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/swift/test/IRGen/builtin_isConcrete.sil:9:17: error: CHECK-LABEL: expected string not found in input
21:53:56 // CHECK-LABEL: define swiftcc i1 @isConcrete_true() {{.*}} {
21:53:56                 ^
21:53:56 <stdin>:1:1: note: scanning from here
21:53:56 ; ModuleID = '-'
21:53:56 ^
21:53:56 <stdin>:6:12: note: possible intended match here
21:53:56 module asm ".section .swift1_autolink_entries,\220x80000000\22"
21:53:56            ^
```

@stephentyrone @gottesmm @atrick any idea as to why IR isn't being emitted in the test suite?

nvzqz replied:
I'm guessing the issue was that I wasn't placing visibility specifiers in the checked function declarations.

nvzqz committed:
Add visibility specifiers to isConcrete IRGen test

nvzqz committed:
Annotate `Swift._isConcrete` with `@_transparent`

Returns `false` otherwise in Debug builds because it is generic.

nvzqz commented:
@swift-ci please smoke test

nvzqz committed:
Add comment on required isConcrete specialization

When calling `Builtin.isConcrete` in SIL, the provided type parameter
should be sufficient.

nvzqz committed:
Add SIL combine test for `Builtin.isConcrete`

nvzqz commented:
@swift-ci please smoke test

stephentyrone commented on test/SILOptimizer/constant_propagation_stdlib.swift:
```diff
@@ -0,0 +1,99 @@
+// CHECK-O:           bb0(
+// CHECK-O:             [[RESULT:%.*]] = integer_literal $Builtin.Int1, -1
+// CHECK-O:             return [[RESULT]]
+// CHECK-O:           } // end sil function '$s27constant_propagation_stdlib26isConcrete_concrete_calleryBi1_AA5MyIntVF'
```
I'm a little bit concerned that the patterns we're matching here include a lot of stuff that isn't directly part of isConcrete; my gut feeling is that these tests are probably going to be pretty fragile. Do any of the people who work on SIL have opinions about this?

atrick replied:
I have a pretty strong opinion that checks against generated SIL should be minimal. They need to be just specific enough not to allow false negatives and no more specific. Rewriting SIL tests whenever SIL changes is a substantial and growing part of our job description.

nvzqz committed:
Generalize constant_propagation_stdlib SIL tests

nvzqz committed:
Remove unused call to `constantFoldIsConcrete`

airspeedswift commented:
@atrick you good to go with this?

airspeedswift commented:
@swift-ci please smoke test

atrick reviewed (approved):
I see the unreachable code in constandFoldBuiltin was removed. And the .sil test cases look reasonable. Thanks.

nvzqz commented:
@swift-ci please test and merge
